File: codes/HellDream/90.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> lists = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
        Arrays.sort(nums);
        generateSubsets(lists, list, 0,nums);
        return lists;
    }
    
    public void generateSubsets(List<List<Integer>> lists, List<Integer> list, int start, int[]nums){
        if(!lists.contains(list))
            lists.add(list);
        if(start==nums.length) return;
        for(int i=start;i<nums.length;i++){
            List<Integer> tmp = new ArrayList<>(list);
            tmp.add(nums[i]);
            generateSubsets(lists,tmp,i+1,nums);
        }
    }
}

File: codes/Garnetwzy/90.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ret;
        if(nums.size() == 0)
            return ret;
        vector<int> cur;
        help(ret, nums,0, cur);
        return ret;
    }
    
    void help(vector<vector<int>>& ret, vector<int>& nums, int start, vector<int>& cur) {
        ret.push_back(cur);
        int i;
        while(start < nums.size()) {
            cur.push_back(nums[start]);
            help(ret, nums, start+1, cur);
            i = start+1;
            while(i < nums.size() && nums[i] == nums[i-1]) {
                i++;
            }
            start = i;
            cur.pop_back();
        }
    }
};